sdks/python/apache_beam/runners/interactive/caching/streaming_cache.py

```diff
@@ -0,0 +1,124 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from __future__ import absolute_import
+
+from apache_beam.portability.api.beam_interactive_api_pb2 import InteractiveStreamHeader
+from apache_beam.portability.api.beam_interactive_api_pb2 import InteractiveStreamRecord
+from apache_beam.portability.api.beam_runner_api_pb2 import TestStreamPayload
+from apache_beam.utils import timestamp
+from apache_beam.utils.timestamp import Timestamp
+
+
+class StreamingCache(object):
+  """Abstraction that holds the logic for reading and writing to cache.
+  """
+  def __init__(self, readers):
+    self._readers = readers
+
+  class Reader(object):
+    """Abstraction that reads from PCollection readers.
+
+    This class is an Abstraction layer over multiple PCollection readers to be
+    used for supplying the Interactive Service with TestStream events.
+
+    This class is also responsible for holding the state of the clock, injecting
+    clock advancement events, and watermark advancement events.
+    """
+    def __init__(self, readers):
+      self._timestamp = Timestamp.of(0)
+      self._readers = {}
+      self._headers = {}
+      readers = [r.read() for r in readers]
+
+      # The header allows for metadata about an entire stream, so that the data
+      # isn't copied per record.
+      for r in readers:
+        header = InteractiveStreamHeader()
```

Comment:
In a different comment, I wrote that I don't know if I want to subclass the PCollection Cache yet. So I assume here that it will be a simple file based cache of strings that I will read row-by-row without any assumption of underlying format. I may change that assumption in the future when the PCollection Cache code is merged.